File: hoffman-agents-python/src/conscious_agent/agent/simple_world.py
from __future__ import annotations

import random
from dataclasses import dataclass, field

from .world_state import WorldState

NATIVE_TOKENS = [
    "cross", "boundary", "arrive", "return", "enter", "leave",
    "different", "same", "familiar", "again", "expect", "notice",
]


@dataclass
class SimpleWorld:
    n_states: int = 10
    seed: int = 42
    tokens: list[str] | None = None

    def __post_init__(self) -> None:
        self._rng = random.Random(self.seed)
        self._token_pool = self.tokens if self.tokens is not None else NATIVE_TOKENS
        self._states: list[list[str]] = []
        self._current_idx = 0
        self._build_states()
        self._transition_matrix = self._build_transition_matrix()

    def _build_states(self) -> None:
        self._states = []
        for i in range(self.n_states):
            length = self._rng.randint(2, 5)
            tokens = self._rng.choices(self._token_pool, k=length)
            self._states.append(tokens)

    def _build_transition_matrix(self) -> list[list[float]]:
        mat = []
        for i in range(self.n_states):
            row = [self._rng.random() for _ in range(self.n_states)]
            total = sum(row)
            row = [v / total for v in row]
            mat.append(row)
        return mat

    def step(self) -> WorldState:
        row = self._transition_matrix[self._current_idx]
        self._current_idx = self._rng.choices(range(self.n_states), weights=row)[0]
        return WorldState.from_sequence("world", list(self._states[self._current_idx]))

    def reset(self) -> None:
        self._current_idx = self._rng.randint(0, self.n_states - 1)

    @property
    def current_state(self) -> list[str]:
        return list(self._states[self._current_idx])

File: hoffman-agents-react-native/src/agent/simple-world.js
const { WorldState } = require('./world-state');

const NATIVE_TOKENS = [
  'cross', 'boundary', 'arrive', 'return', 'enter', 'leave',
  'different', 'same', 'familiar', 'again', 'expect', 'notice',
];

class SimpleWorld {
  constructor({ nStates = 10, seed = 42, tokens = null } = {}) {
    this.nStates = nStates;
    this._rng = seedRandom(seed);
    this._tokenPool = tokens || NATIVE_TOKENS;
    this._states = [];
    this._currentIdx = 0;
    this._buildStates();
    this._transitionMatrix = this._buildTransitionMatrix();
  }

  _buildStates() {
    for (let i = 0; i < this.nStates; i++) {
      const length = this._rng() * 3 + 2;
      const tokens = [];
      for (let j = 0; j < length; j++) {
        tokens.push(this._tokenPool[Math.floor(this._rng() * this._tokenPool.length)]);
      }
      this._states.push(tokens);
    }
  }

  _buildTransitionMatrix() {
    const mat = [];
    for (let i = 0; i < this.nStates; i++) {
      const row = [];
      let total = 0;
      for (let j = 0; j < this.nStates; j++) {
        const v = this._rng();
        row.push(v);
        total += v;
      }
      mat.push(row.map(v => v / total));
    }
    return mat;
  }

  step() {
    const row = this._transitionMatrix[this._currentIdx];
    let r = this._rng(), cumulative = 0;
    for (let j = 0; j < row.length; j++) {
      cumulative += row[j];
      if (r <= cumulative) { this._currentIdx = j; break; }
    }
    return WorldState.fromSequence('world', [...this._states[this._currentIdx]]);
  }

  reset() {
    this._currentIdx = Math.floor(this._rng() * this.nStates);
  }

  get currentState() { return [...this._states[this._currentIdx]]; }
}

function seedRandom(seed) {
  let s = seed;
  return function () {
    s = (s * 1664525 + 1013904223) & 0xFFFFFFFF;
    return (s >>> 0) / 0xFFFFFFFF;
  };
}

module.exports = { SimpleWorld };
